### DISCUSSION TITLE
[SOLVED] Allow selection of local_port for sending UDP packets.

### DISCUSSION THREAD

**Discussion by CabalCrow** [godotengine/godot-proposals#11114] [Engine Core]
## EDIT
Apperantly this already exist and it is done via the ``PackedPeerUDP.bind`` method.

## Suggestion
Currently you can send UDP packets via
```gdscript
	var peer = PacketPeerUDP.new()
	peer.set_dest_address(address, port)
	peer.put_packet(buffer)
```
however those packets would be send with a random local_port. Instead add a way to specify the local port via the ``set_dest_address`` as a 3rd optional argument.

## Why is this important
To enable NAT traversal. It is vital for many NAT traversal techniques to specify the port you send packets from.
This can then be used to improve the internet play of **any** peer to peer game, even without a STUN server.

## Example use case
Two players with ipv6 want to connect over the internet without port forwarding.
Since ipv6 is rarely masks ports, you don't need a STUN server to determine their public_port, as it would coincide with the local_port.
This means the players can connect directly via hole punching in the following way:
1. Player A sends packets to Player B via:
```gdscript
	var peer = PacketPeerUDP.new()
	peer.set_dest_address(player_b_address, player_b_local_port, player_a_local_port)
	peer.put_packet(buffer)
```
2. Player B sends packets back to Player A via:
```gdscript
	var peer = PacketPeerUDP.new()
	peer.set_dest_address(player_a_address, player_a_local_port, player_b_local_port)
	peer.put_packet(buffer)
```
3. This should be enough for the players to successfully establish connection and recieve each other packets. In order words the players have punched thourgh and can now connect and play via the ``player_a_local_port`` and ``player_b_local_port``.

## Possible workarounds
Do not use Godot's PacketPeerUDP, and instead package ``netcat-openbsd`` as a binary and run it from godot to send packets. This is undesireable since trying package and run binaries on different machines and OSes would create a lot of additional problems.

That being said here how **currently** you can achive the same result as the example use case, using the external program ``net-cat-openbsd``.
1. Player A sends packets to Player B via (``nc`` is ``netcat-openbsd``):
```bash
echo "buffer content" | nc -6u player_b_address player_b_local_port -p player_a_local_port
```
2. Player B sends packets back to Player A via:
```bash
echo "buffer content" | nc -6u player_a_address player_a_local_port -p player_b_local_port
```
3. This should be enough for the players to successfully establish connection and recieve each other packets. In order words the players have punched thourgh and can now connect and play via the ``player_a_local_port`` and ``player_b_local_port``.

**Comment by ghost** [comment]
I am not familiar with networking but isn't [bind method](https://docs.godotengine.org/en/latest/classes/class_packetpeerudp.html#class-packetpeerudp-method-bind) what you need?

**Reply by CabalCrow** [reply]
The bind method affects **listening**, not sending of packets.

**Reply by Faless** [reply]
Bind "binds" the socket (see [bind(2)](https://man7.org/linux/man-pages/man2/bind.2.html)) which effectively selects the outgoing port.

If a socket is not bound, the OS selects the outgoing port at random (on Linux in the range 32768 - 60999 by default)

**Comment by Calinou** [comment]
>  Since ipv6 is rarely masks ports, you don't need a STUN server to determine their public_port, as it would coincide with the local_port.

Most ISPs have an IPv6 firewall active by default that blocks all incoming connections at the router level, so this won't work reliably unless the app uses UPnPv6 to open ports.

This firewall is sometimes not controllable on some connection types (mainly mobile connections where you can't open any ports on it, including residential 4G/5G connections).

**Reply by CabalCrow** [reply]
This is why I'm using the NAT traversal technique explained above to punch a hole through the firewall.
The possible workaround I've given at the bottom of the proposal works with IPv6 firewalls.
